File: app/src/main/java/com/dhsdevelopments/potato/clientapi/NotificationTypeAdapter.java
package com.dhsdevelopments.potato.clientapi;

import com.dhsdevelopments.potato.clientapi.notifications.MessageNotification;
import com.dhsdevelopments.potato.clientapi.notifications.PotatoNotification;
import com.dhsdevelopments.potato.clientapi.notifications.StateUpdateNotification;
import com.dhsdevelopments.potato.clientapi.notifications.TypingNotification;
import com.google.gson.*;

import java.lang.reflect.Type;

public class NotificationTypeAdapter implements JsonDeserializer<PotatoNotification>
{
    @Override
    public PotatoNotification deserialize( JsonElement json, Type typeOfT, JsonDeserializationContext context ) throws JsonParseException {
        JsonObject obj = json.getAsJsonObject();
        String type = obj.get( "type" ).getAsString();
        switch( type ) {
            case "m":
                return context.deserialize( obj, MessageNotification.class );
            case "cu":
                return context.deserialize( obj, StateUpdateNotification.class );
            case "type":
                return context.deserialize( obj, TypingNotification.class );
            default:
                return context.deserialize( obj, PotatoNotification.class );
        }
    }
}

File: app/src/main/java/com/dhsdevelopments/potato/clientapi/NotificationTypeAdapter.kt
package com.dhsdevelopments.potato.clientapi

import com.dhsdevelopments.potato.clientapi.notifications.MessageNotification
import com.dhsdevelopments.potato.clientapi.notifications.PotatoNotification
import com.dhsdevelopments.potato.clientapi.notifications.StateUpdateNotification
import com.dhsdevelopments.potato.clientapi.notifications.TypingNotification
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type

class NotificationTypeAdapter : JsonDeserializer<PotatoNotification> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): PotatoNotification {
        val obj = json.asJsonObject
        val type = obj.get("type").asString
        return when (type) {
            "m" -> context.deserialize<PotatoNotification>(obj, MessageNotification::class.java)
            "cu" -> context.deserialize<PotatoNotification>(obj, StateUpdateNotification::class.java)
            "type" -> context.deserialize<PotatoNotification>(obj, TypingNotification::class.java)
            else -> context.deserialize<PotatoNotification>(obj, PotatoNotification::class.java)
        }
    }
}
